### PR TITLE
Add latest-major.minor links and multi-arch support

### PR DESCRIPTION
**Pull request by jupierce**
Running in single arch mode: 
```
16:38 $ mkdir -p /srv/pub/openshift-v4/clients/ocp/4.2.4

16:38 $ ./set-v4-client-latest.sh 4.2.4 ocp all
Can't find x86_64 arch; assuming single arch mode is still in place
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/clients/ocp
'latest' -> '4.2.4'
Overall latest now points to 4.2.4 in /srv/pub/openshift-v4/clients/ocp

16:38 $ ls -l /srv/pub/openshift-v4/clients/ocp
total 4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:38 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:38 latest -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:38 latest-4.2 -> 4.2.4

16:38 $ mkdir -p /srv/pub/openshift-v4/clients/ocp/4.2.5

16:39 $ ./set-v4-client-latest.sh 4.2.5 ocp all
Can't find x86_64 arch; assuming single arch mode is still in place
'latest-4.2' -> '4.2.5'
latest-4.2  now points to 4.2.5 in /srv/pub/openshift-v4/clients/ocp
'latest' -> '4.2.5'
Overall latest now points to 4.2.5 in /srv/pub/openshift-v4/clients/ocp

16:39 $ ls -l /srv/pub/openshift-v4/clients/ocp
total 8
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:38 4.2.4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.2.5
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest -> 4.2.5
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest-4.2 -> 4.2.5

16:39 $ ./set-v4-client-latest.sh 4.2.4 ocp all
Can't find x86_64 arch; assuming single arch mode is still in place
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/clients/ocp
'latest' -> '4.2.4'
Overall latest now points to 4.2.4 in /srv/pub/openshift-v4/clients/ocp

16:39 $ ls -l /srv/pub/openshift-v4/clients/ocp
total 8
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:38 4.2.4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.2.5
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest-4.2 -> 4.2.4

16:39 $ mkdir -p /srv/pub/openshift-v4/clients/ocp/4.3.0

16:39 $ ./set-v4-client-latest.sh 4.3.0 ocp all
Can't find x86_64 arch; assuming single arch mode is still in place
'latest-4.3' -> '4.3.0'
latest-4.3  now points to 4.3.0 in /srv/pub/openshift-v4/clients/ocp
'latest' -> '4.3.0'
Overall latest now points to 4.3.0 in /srv/pub/openshift-v4/clients/ocp

16:39 $ ls -l /srv/pub/openshift-v4/clients/ocp
total 12
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:38 4.2.4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.2.5
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest -> 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest-4.2 -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest-4.3 -> 4.3.0

16:39 $ ./set-v4-client-latest.sh 4.2.5 ocp all
Can't find x86_64 arch; assuming single arch mode is still in place
'latest-4.2' -> '4.2.5'
latest-4.2  now points to 4.2.5 in /srv/pub/openshift-v4/clients/ocp
This is not the highest Y release -- will not set overall latest

16:40 $ ls -l /srv/pub/openshift-v4/clients/ocp
total 12
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:38 4.2.4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.2.5
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:39 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest -> 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:40 latest-4.2 -> 4.2.5
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:39 latest-4.3 -> 4.3.0
```

Running in multi-arch mode:
```
16:43 $ mkdir -p /srv/pub/openshift-v4/s390x/clients/ocp/4.2.4

16:44 $ mkdir -p /srv/pub/openshift-v4/x86_64/clients/ocp/4.2.4

16:44 $ ./set-v4-client-latest.sh 4.2.4 ocp all
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/s390x/clients/ocp
'latest' -> '4.2.4'
Overall latest now points to 4.2.4 in /srv/pub/openshift-v4/s390x/clients/ocp
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/x86_64/clients/ocp
'latest' -> '4.2.4'
Overall latest now points to 4.2.4 in /srv/pub/openshift-v4/x86_64/clients/ocp

16:44 $ ls -l /srv/pub/openshift-v4/x86_64/clients/ocp
total 4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:44 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:44 latest -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:44 latest-4.2 -> 4.2.4

16:44 $ ls -l /srv/pub/openshift-v4/s390x/clients/ocp
total 4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:44 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:44 latest -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:44 latest-4.2 -> 4.2.4

16:44 $ mkdir -p /srv/pub/openshift-v4/x86_64/clients/ocp/4.3.0

16:45 $ ./set-v4-client-latest.sh 4.2.4 ocp any
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/s390x/clients/ocp
'latest' -> '4.2.4'
Overall latest now points to 4.2.4 in /srv/pub/openshift-v4/s390x/clients/ocp
'latest-4.2' -> '4.2.4'
latest-4.2  now points to 4.2.4 in /srv/pub/openshift-v4/x86_64/clients/ocp
This is not the highest Y release -- will not set overall latest

16:45 $ ./set-v4-client-latest.sh 4.3.0 ocp any
Unable to find 4.3.0 directory in /srv/pub/openshift-v4/s390x/clients/ocp ; will not update latest
Failure creating links in /srv/pub/openshift-v4/s390x/clients/ocp
'latest-4.3' -> '4.3.0'
latest-4.3  now points to 4.3.0 in /srv/pub/openshift-v4/x86_64/clients/ocp
'latest' -> '4.3.0'
Overall latest now points to 4.3.0 in /srv/pub/openshift-v4/x86_64/clients/ocp

16:45 $ ls -l /srv/pub/openshift-v4/s390x/clients/ocp
total 4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:44 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:45 latest -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:45 latest-4.2 -> 4.2.4

16:46 $ ls -l /srv/pub/openshift-v4/x86_64/clients/ocp
total 8
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:44 4.2.4
drwxrwxr-x. 2 jupierce jupierce 4096 Nov  8 16:45 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:45 latest -> 4.3.0
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:45 latest-4.2 -> 4.2.4
lrwxrwxrwx. 1 jupierce jupierce    5 Nov  8 16:45 latest-4.3 -> 4.3.0
```

### CODE REVIEW

File: build-scripts/use-mirror/set-v4-client-latest.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+# We are transitioning to multi-arch builds. At the time of this writing, directories like
+# ${BASE_DIR}/${arch}/clients don't exist -- only ${BASE_DIR}/clients does. This script attempts to account
+# for the present and future. if x86_64 doesn't exist yet, we assume clients/ contains the x864_64 payload.
+
+if [[ ! -d "${BASE_DIR}/x86_64" ]]; then
+    echo "Can't find x86_64 arch; assuming single arch mode is still in place"
+    create_latest_links ${BASE_DIR}/clients/${CLIENT_TYPE}
+else
+
+    # Otherwise, multi-arch is enabled. Run the links logic for each architecture passed in.
```
Comment (vfreex) [inline build-scripts/use-mirror/set-v4-client-latest.sh]: When multi-arch is enabled, what should we do with the old "single arch" path? Should we create a symlink to the `x86_64` path?

File: build-scripts/use-mirror/set-v4-client-latest.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        if [[ ! -d "${target_dir}" ]]; then
+            echo "Unable to find clients directory under: ${target_dir}"
+            continue
+        fi
+
+        if ! create_latest_links "${target_dir}"; then
+            echo "Failure creating links in ${target_dir}"
+            if [[ "${MODE}" != "any" ]]; then
+                echo "This was a required operation; exiting with failure"
+                exit 1
```
Comment (vfreex) [inline build-scripts/use-mirror/set-v4-client-latest.sh]: Should `/usr/local/bin/push.pub.sh` be called if this happens? Considering some symlinks might be already changed on use-mirror, those change could be pushed in succeeding runs.

File: build-scripts/use-mirror/set-v4-client-latest.sh [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    ln -svfn ${RELEASE} latest
+    echo "Overall latest now points to ${RELEASE} in ${client_base_dir}"
+    return 0
+}
+
+
+# We are transitioning to multi-arch builds. At the time of this writing, directories like
+# ${BASE_DIR}/${arch}/clients don't exist -- only ${BASE_DIR}/clients does. This script attempts to account
+# for the present and future. if x86_64 doesn't exist yet, we assume clients/ contains the x864_64 payload.
```
Comment (vfreex) [inline build-scripts/use-mirror/set-v4-client-latest.sh]: `s/x864_64/x86_64`

File: build-scripts/use-mirror/set-v4-client-latest.sh (hunk 1 above)
Comment (jupierce) [reply build-scripts/use-mirror/set-v4-client-latest.sh]: Yes - they should be turned into symlinks directed into the x86_64 subdirectories for backwards compatibility.

File: build-scripts/use-mirror/set-v4-client-latest.sh (hunk 2 above)
Comment (jupierce) [reply build-scripts/use-mirror/set-v4-client-latest.sh]: That's a good point and I had not considered it. But I think I would prefer to give a human a chance to try to figure out what went wrong before pushing that unexpected state to the world. It still might, if something else does a push, but it at least gives us a window to investigate.

File: build-scripts/use-mirror/set-v4-client-latest.sh [hunk 4]
```diff
@@ -0,0 +1,107 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (jupierce) [reply build-scripts/use-mirror/set-v4-client-latest.sh]: hmm.. I did do a move, but I guess the rename & dir change mixed up git enough it couldn't figure it out.

File: build-scripts/use-mirror/set-v4-client-latest.sh (hunk 1 above)
Comment (vfreex) [reply build-scripts/use-mirror/set-v4-client-latest.sh]: OK.

File: build-scripts/use-mirror/set-v4-client-latest.sh (hunk 2 above)
Comment (vfreex) [reply build-scripts/use-mirror/set-v4-client-latest.sh]: Makes sense. Thanks.

**Pull request merged**